File: src/shesasyncer/lyrics/normalize.py
import re
import unicodedata


def normalize_lyric(text: str) -> str:
    """Normalize text for matching without changing the trusted lyric itself."""
    text = unicodedata.normalize("NFKC", text).casefold()
    return re.sub(r"[^\w\u00C0-\uFFFF]+", "", text, flags=re.UNICODE)


def lyric_tokens(text: str) -> list[str]:
    return re.findall(r"\w+", unicodedata.normalize("NFKC", text).casefold(), flags=re.UNICODE)

File: src/shesasyncer/lyrics/normalize.js
export function normalizeLyric(text) {
  return String(text).normalize('NFKC').toLocaleLowerCase().replace(/[^\p{L}\p{N}_]+/gu, '');
}

export function lyricTokens(text) {
  return String(text).normalize('NFKC').toLocaleLowerCase().match(/[\p{L}\p{N}_]+/gu) ?? [];
}
